crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+    matrix: &Matrix,
+    v: &PatStack,
+) -> MatchCheckResult<Usefulness> {
+    if v.is_empty() {
+        let result = if matrix.is_empty() { Usefulness::Useful } else { Usefulness::NotUseful };
+
+        return Ok(result);
+    }
+
+    if let Pat::Or(pat_ids) = v.head().as_pat(cx) {
+        let mut found_unimplemented = false;
+        let any_useful = pat_ids.iter().any(|&pat_id| {
+            let v = PatStack::from_pattern(pat_id);
+
+            match is_useful(cx, matrix, &v) {
+                Ok(Usefulness::Useful) => true,
+                Ok(Usefulness::NotUseful) => false,
+                _ => {
+                    found_unimplemented = true;
+                    false
+                }
+            }
+        });
+
+        return if any_useful {
+            Ok(Usefulness::Useful)
+        } else if found_unimplemented {
+            Err(MatchCheckNotImplemented)
+        } else {
+            Ok(Usefulness::NotUseful)
+        };
+    }
+
+    if let Some(constructor) = pat_constructor(cx, v.head())? {
+        let matrix = matrix.specialize_constructor(&cx, &constructor)?;
+        let v = v
+            .specialize_constructor(&cx, &constructor)?
+            .expect("we know this can't fail because we get the constructor from `v.head()` above");
+
+        is_useful(&cx, &matrix, &v)
+    } else {
+        // expanding wildcard
+        let mut used_constructors: Vec<Constructor> = vec![];
+        for pat in matrix.heads() {
+            if let Some(constructor) = pat_constructor(cx, pat)? {
+                used_constructors.push(constructor);
+            }
+        }
+
+        // We assume here that the first constructor is the "correct" type. Since we
+        // only care about the "type" of the constructor (i.e. if it is a bool we
+        // don't care about the value), this assumption should be valid as long as
+        // the match statement is well formed. We currently uphold this invariant by
+        // filtering match arms before calling `is_useful`, only passing in match arms
+        // whose type matches the type of the match expression.
+        match &used_constructors.first() {
+            Some(constructor) if all_constructors_covered(&cx, constructor, &used_constructors) => {
+                // If all constructors are covered, then we need to consider whether
+                // any values are covered by this wildcard.
+                //
+                // For example, with matrix '[[Some(true)], [None]]', all
+                // constructors are covered (`Some`/`None`), so we need
+                // to perform specialization to see that our wildcard will cover
+                // the `Some(false)` case.
+                let mut constructor = None;
+                for pat in matrix.heads() {
+                    if let Some(c) = pat_constructor(cx, pat)? {
+                        constructor = Some(c);
+                        break;
+                    }
+                }
```

Comment:
this seems like it's just getting `used_constructors.first()` again? :thinking:

Reply:
Oops that was silly. Thanks for pointing this out.